File: 프로그래머스/3/43163. 단어 변환/단어 변환.swift
import Foundation

func canChange(from: String, to: String, done: [String]) -> Bool {
    if done.contains(to) { return false }
    
    var diff = 0
    let f = Array(from)
    let t = Array(to)

    for i in 0..<f.count {
        if f[i] != t[i]  { diff += 1 }
    }
    
    return diff == 1
}

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    var queue: [[String]] = []
    var ans = 0
    
    queue.append([begin])
    while !queue.isEmpty {
        var pop = queue.removeFirst()
        guard let last = pop.last else { break }

        if last == target { 
            ans = pop.count - 1
            break
        }
        for word in words {
            if canChange(from: last, to: word, done: pop) {
                let newArr = pop + [word]
                queue.append(newArr)
            }
        }
    }

    return ans
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const used = [];
    const stack = [[begin, 0]];
    
    if (!words.includes(target)) return 0;
    
    while (stack.length) {
        const [word, count] = stack.shift();
        // 동일하면 종료
        if (word === target) return count;
        
        if (used.includes(word)) continue;
        used.push(word);
        
        for (let w of words) {
            // 1개 차이 나면
            if (isOneChange(word, w)) {
                stack.push([w, count + 1]);
            }
        }
    }
    
    return 0;
}

function isOneChange(cur, next) {
    let count = 0;
    
    for (let i = 0; i < cur.length; i++) {
        if (cur[i] !== next[i]) count++;
        if (count > 1) return false;
    }
    
    return count === 1 ? true : false;
}